### ISSUE TITLE
TypeError: Item of collection User.ips must be an instance of User_IP. Got: '1'

### ISSUE THREAD

**Issue by interneteur** [ponyorm/pony#130]
Having a issue where I'm trying to create a new user registration.  I'm trying to store the user ip in a Set

Everything works fine execept when I try to store their IP address.

My models looks like this
```
class User(db.Entity):
    username = Required(unicode, 20, unique=True)
    email = Required(unicode, 255, unique=True)
    password_hash = Optional(unicode, 255)
    main = Optional("User", reverse="alts")
    alts = Set("User", reverse="main")
    roles = Set("Role")
    permissions = Set("Permission")
    profile = Optional("UserProfile")
    is_confirmed = Required(bool, default=0)
    is_active = Required(bool, default=0)
    is_main = Required(bool, default=0)
    confirmed_at = Required(datetime, default=lambda: datetime.now())
    last_login = Required(datetime, default=lambda: datetime.now())
    current_login = Required(datetime, default=lambda: datetime.now())
    ips = Set("User_IP")
    favorites = Optional("Favorite")
    upvotes = Optional("Game_Upvotes")

    def hash_password(self, password):
        self.password_hash = pwd_context.encrypt(password)

    def verify_password(self, password):
        return pwd_context.verify(password, self.password_hash)


class User_IP(db.Entity):
    user = Required(User)
    ip_address = Required(IP_Address)
    date_used = Required(datetime)
    PrimaryKey(user, ip_address)

class IP_Address(db.Entity):
    address = Required(unicode, 45, default='127.0.0.1')
    users = Set("User_IP")
```
My Function like this

```
@app.route('/register', methods=['GET', 'POST'])
@pny.db_session
def register():
    form = RegistrationForm(request.form)
    print request.environ['REMOTE_ADDR']
    if request.method == 'POST' and form.validate():
        user = User(username=form.username.data, email=form.email.data, ips = request.environ['REMOTE_ADDR'])
        user.hash_password(form.password.data)

        flash('Thanks for registering')
        return redirect(url_for('register'))
    return render_template('register.html', form=form)
```

And when I sign up it gives me this error

```
TypeError: Item of collection User.ips must be an instance of User_IP. Got: '1'
````

Any help would be apprecicated

**Comment by interneteur** [comment]
I figured it out, was my bad, didn't realize I needed to declare the sets as well.

```

@app.route('/register', methods=['GET', 'POST'])
@pny.db_session
def register():
    form = RegistrationForm(request.form)
    print request.environ['REMOTE_ADDR']
    if request.method == 'POST' and form.validate():
        user = User(username=form.username.data, email=form.email.data)
        user.hash_password(form.password.data)
        ip = IP_Address(address = request.environ['REMOTE_ADDR'])
        User_IP(user = user, ip_address = ip, date_used = datetime.now())
        flash('Thanks for registering')
        return redirect(url_for('register'))
    return render_template('register.html', form=form)
```

**Comment by amalashkevich** [comment]
The `User` entity has the attribue `ips` defined as `ips = Set("User_IP")`, so Pony expects you to assign a set of `User_IP` instances to this attr. When you are trying to assign a string, you get the exception.
In order to fix this, you have to replace the line

`user = User(username=form.username.data, email=form.email.data, ips = request.environ['REMOTE_ADDR'])`

to these lines:

```python
    user = User(username=form.username.data, email=form.email.data)
    remote_address = request.environ['REMOTE_ADDR']
    ip_address = IP_Address.get(address=remote_address)
    if ip_address is None:
        ip_address = IP_Address(address=remote_address)
    user.ips.create(ip_address=ip_address, date_used=datetime.now())
```

I split getting and creating `IP_Address` instance for avoiding duplicates in the database. Also you can add the `unique=True` to the `address` attribute in the `IP_Address` entity definition for checking for duplicates on the database level.

**Comment by LaundroMat** [comment]
Sorry for resurrecting this, but am I right in supposing that if I have this model:

    class Organisation(db.Entity):
        name = Required(str)
        members = Set('Person')

    class Person(db.Entity):
        email = Required(str)
        organisation = Optional(Organisation)

I *can* create a Person using

    org = Organisation(name="ACME Inc")
    person = Person(email="fake@fake.org", organisation=org.id)

but I cannot add Person instances to an Organisation using the Person id's?

    person = Person(email="member@acme.org")
    org = Organisation(name="ACME Inc", members=[person.id])
    # -> Error TypeError: Item of collection Organisation.members must be an instance of Person Got: 1

In other words, I must first retrieve an instance of the person from the database (or create the instance of Person, as in the example above), before I can use it in a Set() when creating an instance of an Organisation.
